#!/bin/bash

# Change the paths as necessary
# DATA_DIR is the location of raw fastq files
# PROJECT_DIR is the main directory for this analysis
# ALL_FILE_LIST is the list of renamed sample fastq files

export PROJECT_DIR="/home/smalkaram/Projects/alwayMRNA20170516143625"
export DATA_DIR="/home/smalkaram/Projects/alwayMRNA20170516143625/raw_data/file_links"
export ALL_FILE_LIST="/home/smalkaram/Projects/alwayMRNA20170516143625/ALL_FILE_LIST"

module load bio/processing/Fastqc/0.11.5
module load pkg/perl5/5.22.1
module load pkg/anaconda2/4.3.1
module load bio/alignment/mirdeep/2.0.0.8

# Change the following paths as needed. 
# ARF and READS are generated by mapperMOD.pl
# requies miRBase release21 files for mmu, rno and cgr organism
READS="../reads2.fa"
GENOME="/data/db/GenCode/GRCm38p5/GRCm38.primary_assembly.genome.fa"
ARF="../reads_vs_genome2.arf"
MATURE="/data/db/miRNA/miRBase/release21/mmu_mature.fa"
RELATED="/data/db/miRNA/miRBase/release21/rno_cgr_mature.fa"
PRECURSORS="/data/db/miRNA/miRBase/release21/mmu_hairpin.fa"

miRDeep2.pl $READS $GENOME $ARF $MATURE $RELATED $PRECURSORS -t Mouse 2>report.log
# Sridhar A Malkaram (smalkaram@wvstateu.edu)
# Last modified on: 12/30/2017
